src/util/fifo.h

```diff
@@ -1,69 +1,116 @@
 #pragma once
 
-#include "pa_ringbuffer.h"
+#include <cstdint>
 
 #include "util/class.h"
 #include "util/math.h"
 
+using ring_buffer_size_t = uint32_t;
+
 template <class DataType>
 class FIFO {
   public:
-    explicit FIFO(int size)
-            : m_data(roundUpToPowerOf2(size)) {
-        // If we can't represent the next higher power of 2 then bail.
-        if (m_data.size() == 0) {
-            return;
-        }
-        PaUtil_InitializeRingBuffer(&m_ringBuffer,
-                static_cast<ring_buffer_size_t>(sizeof(DataType)),
-                static_cast<ring_buffer_size_t>(m_data.size()),
-                m_data.data());
+    explicit FIFO(uint32_t size)
+            : m_size{roundUpToPowerOf2(size)},
+              m_mask{m_size - 1},
+              m_data(m_size),
+              m_writeIndex{0},
+              m_readIndex{0} {
     }
-    virtual ~FIFO() {
+    uint32_t readAvailable() const {
+        const uint32_t readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const uint32_t writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        return writeIndex - readIndex;
     }
-    int readAvailable() const {
-        return PaUtil_GetRingBufferReadAvailable(&m_ringBuffer);
+    uint32_t writeAvailable() const {
+        const uint32_t readIndex = m_readIndex.load(std::memory_order_acquire);
+        const uint32_t writeIndex = m_writeIndex.load(std::memory_order_relaxed);
+        return m_size - (writeIndex - readIndex);
     }
-    int writeAvailable() const {
-        return PaUtil_GetRingBufferWriteAvailable(&m_ringBuffer);
+    uint32_t read(DataType* pData, uint32_t count) {
+        const uint32_t readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const uint32_t writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        const uint32_t available = writeIndex - readIndex;
+        count = std::min(available, count);
+        const uint32_t n = std::min(m_size - (readIndex & m_mask), count);
+        std::memcpy(pData, m_data.data() + (readIndex & m_mask), n * sizeof(DataType));
+        std::memcpy(pData + n, m_data.data(), (count - n) * sizeof(DataType));
```

Comment:
Pretty sure this requires https://en.cppreference.com/w/cpp/types/is_trivially_copyable, right?

Reply:
It does.

Reply:
right, so then we should either consider adding a `static_assert` for that or use the appropriate more generic algorithm (`std::copy`?).

Reply:
Yes, I’ll change it to std::copy.  Old habits die hard and I often notice that you are not plagued by the baggage of over 30 years of hardcore C and primitive C++ 😅

Reply:
std::copy is slow:
https://github.com/mixxxdj/mixxx/blob/2397016ea899185207e03912773ec43553d0c6f2/src/util/sample.h#L62

We want here a replacement for the original implementation, that only works with trivial types. For C++ objects, we have already an alternative.

Reply:
The speed here depends entirely on the stdlib implementation. Those benchmarks have been made 9 years ago, I doubt they accurately reflect the current state of things. I think a re-evaluation is in order before making a decision based on these numbers.

Reply:
fyi, A [quick and dirty quickbench](https://quick-bench.com/q/yS9ve_ZtfA3UUV0CaVPuiJY9Ack) reveals that there is zero difference (apart from the zero-sized case, in which case `std::copy` wins).

Reply:
Nice. Thank you.

Reply:
Yes, I agree with Niko, my tests show the same. I find the for loop in [mixxx/src/util/sample.h](https://github.com/mixxxdj/mixxx/blob/2397016ea899185207e03912773ec43553d0c6f2/src/util/sample.h#L62) more dubious!

Reply:
using std::copy, please check and resolve.